client/client.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2020 - for information on the respective copyright owner
+// see the NOTICE file and/or the repository at
+// https://github.com/direct-state-transfer/dst-go
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/channel/persistence/keyvalue"
+	"perun.network/go-perun/client"
+	"perun.network/go-perun/peer"
+	"perun.network/go-perun/pkg/sortedkv/leveldb"
+
+	"github.com/direct-state-transfer/dst-go"
+	"github.com/direct-state-transfer/dst-go/blockchain/ethereum"
+)
+
+// Client is a wrapper type around the state channel client implementation from go-perun.
+// It also manages the lifecycle of listener and handler go-routines.
+type Client struct {
+	*client.Client
+
+	wg *sync.WaitGroup
+}
+
+// NewEthereumPaymentClient initializes a two party, ethereum payment channel client for the given user.
+// It establishes a connection to the blockchain and verifies the integrity of contracts at the given address.
+// It uses the comm backend to initialize adapters for off-chain communication network.
+func NewEthereumPaymentClient(cfg Config, user dst.User, comm dst.CommBackend) (*Client, error) {
+	dialer, listener, err := initComm(comm, user.CommAddr)
+	if err != nil {
+		return nil, err
+	}
+	funder, adjudicator, err := connectToChain(cfg.Chain, user.OnChain)
+	if err != nil {
+		return nil, err
+	}
+	// Only off-chain account is unlocked. Accounts for the participant addresses
+	// will be unlocked by the client when required.
+	offChainAcc, err := user.OffChain.Wallet.Unlock(user.OffChain.Addr)
+	if err != nil {
+		return nil, errors.WithMessage(err, "off-chain address")
```

Comment:
I have revisited all the error messages and structured it such that, additional message is added only when some meaningful information can be added.

In this particular case, the `Unlock` function will add a message  "unlocking" and at this point, only "off-chain address" will be added. So that, the complete error message would be  "off-chain address: unlocking".